pyi codegen refactor - no need to group python signatures by overload name

bdhirsh opened the pull request:
Now that all of the byte-for-byte hacks are removed in the pyi codegen, there's no reason for the codegen to group pyi signature overloads together. I updated the logic in `gen_pyi` that computes signatures (`generate_type_hints()` and `generate_named_tuples()`) to operate per individual `PythonSignatureGroup`

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49057 pyi codegen refactor - no need to group python signatures by overload name**
* #49056 pyi codegen - removing byte-for-byte-compatibility hacks (sorting overloads)
* #49055 pyi codegen - removing byte-for-byte compatibility hacks
* #49054 pyi cleanup

Differential Revision: [D25410849](https://our.internmc.facebook.com/intern/diff/D25410849)

bdhirsh committed:
pyi codegen refactor - no need to group python signatures by overload name

[ghstack-poisoned]

bdhirsh committed:
Update on "pyi codegen refactor - no need to group python signatures by overload name"

Now that all of the byte-for-byte hacks are removed in the pyi codegen, there's no reason for the codegen to group pyi signature overloads together. I updated the logic in `gen_pyi` that computes signatures (`generate_type_hints()` and _generate_named_tuples()`) to operate per individual `PythonSignatureGroup`


Differential Revision: [D25410849](https://our.internmc.facebook.com/intern/diff/D25410849)

[ghstack-poisoned]

bdhirsh committed:
Update on "pyi codegen refactor - no need to group python signatures by overload name"

Now that all of the byte-for-byte hacks are removed in the pyi codegen, there's no reason for the codegen to group pyi signature overloads together. I updated the logic in `gen_pyi` that computes signatures (`generate_type_hints()` and _generate_named_tuples()`) to operate per individual `PythonSignatureGroup`


Differential Revision: [D25410849](https://our.internmc.facebook.com/intern/diff/D25410849)

[ghstack-poisoned]

bdhirsh committed:
Update on "pyi codegen refactor - no need to group python signatures by overload name"

Now that all of the byte-for-byte hacks are removed in the pyi codegen, there's no reason for the codegen to group pyi signature overloads together. I updated the logic in `gen_pyi` that computes signatures (`generate_type_hints()` and _generate_named_tuples()`) to operate per individual `PythonSignatureGroup`


Differential Revision: [D25410849](https://our.internmc.facebook.com/intern/diff/D25410849)

[ghstack-poisoned]

ezyang commented on tools/pyi/gen_pyi.py:
```diff
@@ -176,54 +176,44 @@ def sig_for_ops(opname: str) -> List[str]:
-    members of the Tensor class (true) or functions in the
-    `torch` namespace (default, false).
-    """
+def generate_named_tuples(sig_group: PythonSignatureGroup, namedtuples: Dict[str, str]) -> None:
```
nit: if you can, try to avoid nonlocal mutation like what's going on in this function. The moment you start have to start reasoning across function calls to figure out what is mutating, it is just a bad look. You could for example, maintain the old structure and have it return a `Tuple[str, str]`, and then assign into `namedtuples` at the call site, keeping the mutation local.

bdhirsh replied:
The original codegen had an assertion against global state that was worth keeping in (for every named_tuple we encounter, check that its definition doesn't collide with an existing one).

Totally agreed with what you're saying- I'm fixing that by getting rid of `generate_named_tuples` entirely and moving the logic to the call-site.

bdhirsh committed:
Update on "pyi codegen refactor - no need to group python signatures by overload name"

Now that all of the byte-for-byte hacks are removed in the pyi codegen, there's no reason for the codegen to group pyi signature overloads together. I updated the logic in `gen_pyi` that computes signatures (`generate_type_hints()` and `generate_named_tuples()`) to operate per individual `PythonSignatureGroup`


Differential Revision: [D25410849](https://our.internmc.facebook.com/intern/diff/D25410849)

[ghstack-poisoned]

bdhirsh committed:
Update on "pyi codegen refactor - no need to group python signatures by overload name"

Now that all of the byte-for-byte hacks are removed in the pyi codegen, there's no reason for the codegen to group pyi signature overloads together. I updated the logic in `gen_pyi` that computes signatures (`generate_type_hints()` and `generate_named_tuples()`) to operate per individual `PythonSignatureGroup`


Differential Revision: [D25410849](https://our.internmc.facebook.com/intern/diff/D25410849)

[ghstack-poisoned]

bdhirsh committed:
Update on "pyi codegen refactor - no need to group python signatures by overload name"

Now that all of the byte-for-byte hacks are removed in the pyi codegen, there's no reason for the codegen to group pyi signature overloads together. I updated the logic in `gen_pyi` that computes signatures (`generate_type_hints()` and `generate_named_tuples()`) to operate per individual `PythonSignatureGroup`


Differential Revision: [D25410849](https://our.internmc.facebook.com/intern/diff/D25410849)

[ghstack-poisoned]
